frontend/src/components/Modal.tsx

```diff
@@ -0,0 +1,223 @@
+import React, { useRef, useState, useEffect } from "react";
+import styles from "../css/Modal.module.css";
+import Alert from "./Alert";
+import type { AllUserData } from "../types";
+import {
+    blockUser,
+    getAllData,
+    getInterestName,
+    sendFriendRequest,
+    unblockUser,
+} from "../utils";
+import { useUser } from "../contexts/UserContext";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import {
+    faUserCheck,
+    faUserXmark,
+    faCircleArrowRight,
+    faUserShield,
+    faShieldHalved,
+} from "@fortawesome/free-solid-svg-icons";
+
+interface ModalProps {
+    userData: AllUserData | null;
+    setUserData: React.Dispatch<React.SetStateAction<AllUserData | null>>;
+}
+
+/**
+ *
+ * @param userData data to show in modal; null when modal is hidden
+ * @param setUserData state function to update userData; used to hide modal
+ * @returns A modal displaying a user's data overlaid on the content below
+ */
+const Modal = ({ userData, setUserData }: ModalProps) => {
+    // reference to the overlay div itself
+    const overlayRef = useRef<HTMLDivElement | null>(null);
+
+    // closes modal if the overlay itself is clicked, not its contents
+    const handleOverlayClick = (event: React.MouseEvent) => {
+        if (event.target === overlayRef.current) {
+            setUserData(null);
+        }
+    };
+
+    // the logged-in user
+    const { user } = useUser();
+
+    // the full user table data for the logged-in user
+    const [currentUserData, setCurrentUserData] = useState<AllUserData | null>(
+        null,
+    );
+
+    const loadCurrentUserData = async () => {
+        if (user) {
+            const data = await getAllData(user.id);
+            setCurrentUserData(data);
+        }
+    };
+
+    // once user is loaded from context, get user's data
+    useEffect(() => {
+        loadCurrentUserData();
+    }, [user]);
+
+    // message to show in alert (null when alert is not showing)
+    const [alertText, setAlertText] = useState<string | null>(null);
+
+    // try to send a friend request when button is clicked
+    const handleFriendClick = () => {
+        if (userData) {
+            sendFriendRequest(userData.id).then((success) =>
+                setAlertText(
+                    success
+                        ? "Request sent."
+                        : "A friend request between you already exists.",
+                ),
+            );
+        }
+    };
+
+    // block user when button is clicked
+    const handleBlockClick = async () => {
+        if (userData) {
+            await blockUser(userData.id);
+            setAlertText("Successfully blocked user.");
```

Comment:
One other thing you could think about is if you want to check for the success of some of those API requests. Your code assumes success here (and in a few other places) without looking at the http response code (which your backend should ideally provide). Even a simple check that confirms an http success type response from the API and a generic error message in case of an error.